### ISSUE TITLE
crengine crash tries loading deleted font

### ISSUE THREAD

**Issue by yparitcher** [koreader/koreader#6274]
* KOReader version: master
* Device: any

#### Issue

It looks like the font cache is only loaded on the first crengine opening and reused for every new doc. therefore if a font is deleted crengine still tries accessing it and crashes.
This may be out of scope as what is a user doing deleting fonts from within koreader :)

I came across this while deleting different font variations from one family as koreader decided to pick the bold by default.

as a side point can someone explain to me how crengine decides which variant of a font to use? (bold, extra-bold, thin, light, regular)

#### Steps to reproduce

open a document & set a font
go to FM & delete font
Reopen the book without closing Koreader and crash

This does not happen if koreader is closed and reopened

##### `crash.log` (if applicable)
```
06/16/20-22:19:32 DEBUG CreDocument: loading document...
*** Segmentation fault
 
Backtrace:
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_Z12setNodeStyleP8ldomNode9LVFastRefI17css_style_rec_tagE18LVProtectedFastRefI6LVFontE+0x805)[0x7fed6de550af]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN8ldomNode13initNodeStyleEv+0x281)[0x7fed6dd6aa69]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN17ldomElementWriter11onBodyEnterEv+0x40)[0x7fed6dd75974]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN18ldomDocumentWriter9OnTagBodyEv+0x7c)[0x7fed6dd75b68]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN24ldomDocumentWriterFilter9OnTagBodyEv+0x10)[0x7fed6dd75eca]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN11LVXMLParser5ParseEv+0x7ea)[0x7fed6dda001c]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN12LVHTMLParser5ParseEv+0x9)[0x7fed6dda061f]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN9LVDocView13ParseDocumentEv+0x779)[0x7fed6de3b793]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN9LVDocView12LoadDocumentE9LVFastRefI8LVStreamEb+0x175d)[0x7fed6de3d619]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN9LVDocView12LoadDocumentEPKwb+0x915)[0x7fed6de3f24d]
/AUR/koreader/koreader-emulator-x86_64-pc-linux-gnu-debug/koreader/./libs/libcrengine.so(_ZN9LVDocView12LoadDocumentEPKcb+0x80)[0x7fed6de3f554]
./libs/libkoreader-cre.so(+0xab1c)[0x7fed7c31ab1c]
./luajit(+0x22e85)[0x563e0a0fce85]
./luajit(+0x23cd5)[0x563e0a0fdcd5]
./luajit(lua_pcall+0x5a)[0x563e0a0ec6d5]
./luajit(+0x894f)[0x563e0a0e294f]
./luajit(+0x91d7)[0x563e0a0e31d7]
./luajit(+0x98df)[0x563e0a0e38df]
./luajit(+0x22e85)[0x563e0a0fce85]
./luajit(lua_cpcall+0x23)[0x563e0a0ec728]
./luajit(main+0x42)[0x563e0a0e398b]
/usr/lib/libc.so.6(__libc_start_main+0xf2)[0x7fed9231b002]
./luajit(_start+0x2e)[0x563e0a0e262e]
```

**Comment by NiLuJe** [comment]
IIRC, CRe pretty much uses the first style it iterates over, and then starts rejecting fonts from similar families once it's got what it thinks are regular/bold/italic/bolditalic. The validation and matching is minimal and... not pretty ^^.

We've broached the topic a couple times over the years, IIRC.

If you look for the warning printed when CRe rejects Kindle fonts because they already match a cached family, you should be around the right place ;).

**Comment by NiLuJe** [comment]
(And, yeah, I concur with your analysis, FWIW, I updated my Bitter copy after the discussion in #6272, and shit went wonky until the next restart ^^).

**Comment by poire-z** [comment]
Yes, font (and hyphenation) initialization happens only once, the first time we need crengine:
https://github.com/koreader/koreader/blob/5aab341ed0d54e43385c219d4c6f8af9f19a6088/frontend/document/credocument.lua#L96-L118

The FontMan is global and keeps the list of font we gave it here. Given that:
- this phase may take some time, parsing/checking/making fonts and fonts data structures
- checking each time a font is requested and associated to a node if that font still exists would also take time

I'd consider this issue a bit ouf-of-scope edge-casy user-fault, and that it's better to avoid the overhead of handling it.

Some cases I haven't checked (if someone wants to, please do) is:

- open a book large enough so it will uses a cache, change your main font, check it applies to the text, exit KOReader, check there is a cache file
- remove that font
- launch KOReader, does it crash ? (I think it won't as frontend code would give it another font and make the cache invalid and cause a re-rendering).

Variation:
- have a book use some (non embedded) font via a `p { font-family: SomeFont; }`
- open it, set AnotherFont as the main font via the UI menu, verify it does not apply and SomeFont is still used
- exit KOReader, check there is a cache file, remove that SomeFont
- launch KOReader (it may still want to use SomeFont as it was saved to the cache - although I think a hash of the whole list of provided fonts is saved too, and would make the cache invalid then).


>> as a side point can someone explain to me how crengine decides which variant of a font to use? (bold, extra-bold, thin, light, regular)

> IIRC, CRe pretty much uses the first style it iterates over, and then starts rejecting fonts from similar families once it's got what it thinks are regular/bold/italic/bolditalic

Yes, that's right. Some more details at https://github.com/koreader/koreader/issues/4174#issuecomment-414349293

I'm quite fine with the current behaviour: it allows me to just put the Medium and Bold (and their italics) to have Medium used instead of the Regular (when I find it too thin), if I don't put the Regular on the device. Of course, I have to exit KOReader/mount/unmount/relaunch to test all that (and avoid a crash :) - but having some UI to handle all that would be quite complicated.

**Comment by NiLuJe** [comment]
Pssh, scp FTW! :)

**Comment by yparitcher** [comment]
> The FontMan is global and keeps the list of font we gave it here. Given that:
> 
>     * this phase may take some time, parsing/checking/making fonts and fonts data structures
> 
>     * checking each time a font is requested and associated to a node if that font still exists would also take time
> 
> 
> I'd consider this issue a bit ouf-of-scope edge-casy user-fault, and that it's better to avoid the overhead of handling it.
> 

I agree, i would rather fast loading than preventing this crash, I just wasnt sure if there was an easy way around it.

> > > as a side point can someone explain to me how crengine decides which variant of a font to use? (bold, extra-bold, thin, light, regular)
> 
> > IIRC, CRe pretty much uses the first style it iterates over, and then starts rejecting fonts from similar families once it's got what it thinks are regular/bold/italic/bolditalic
> 
> Yes, that's right. Some more details at [#4174 (comment)](https://github.com/koreader/koreader/issues/4174#issuecomment-414349293)

Thanks
